OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -39,6 +43,203 @@ static double UTOL = (double)SimTK::Eps*1e2;
+            : _mX(mX), _mY(mY), _x0(x0), _x1(x1), _y0(y0), _y1(y1),
+              _dydx0(dydx0), _dydx1(dydx1), _computeIntegral(computeIntegral),
+              _intx0x1(intx0x1) {}
+
+    SimTK::Matrix _mX;
+    SimTK::Matrix _mY;
+
+    double _x0;
+    double _x1;
+    double _y0;
+    double _y1;
+    double _dydx0;
+    double _dydx1;
+    bool _computeIntegral;
+    bool _intx0x1;
+};
+
+bool operator==(
+    const SimTK::Matrix &lhs,
+    const SimTK::Matrix &rhs
+) {
+    if (lhs.nrow() != rhs.nrow() || lhs.ncol() != rhs.ncol()) {
+        return false;
+    }
+    for (int r = 0; r < lhs.nrow(); ++r)
+        for (int c = 0; c < lhs.ncol(); ++c)
+            if (lhs.row(r)[c] != rhs.row(r)[c]) return false;
+    return true;
+}
+
+bool operator==(
+    const SmoothSegmentedFunctionParameters &lhs,
+    const SmoothSegmentedFunctionParameters &rhs
+) {
+    return
+    lhs._x0 == rhs._x0 &&
+    lhs._x1 == rhs._x1 &&
+    lhs._y0 == rhs._y0 &&
+    lhs._y1 == rhs._y1 &&
+    lhs._dydx0 == rhs._dydx0 &&
+    lhs._dydx1 == rhs._dydx1 &&
+    lhs._computeIntegral == rhs._computeIntegral &&
+    lhs._intx0x1 == rhs._intx0x1 &&
+    lhs._mX == rhs._mX &&
+    lhs._mY == rhs._mY;
+}
+
+//=============================================================================
+// HASHING OF PARAMETERS
+//=============================================================================
+
+template<typename T>
+inline std::size_t HashCombine(std::size_t seed, T const &v) {
+    std::hash<T> hasher;
+    return seed ^ (hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2));
+}
+
+template<typename T>
+inline std::size_t HashOf(T const &v) {
+    return std::hash<T>{}(v);
+}
+
+template<typename T, typename... Others>
+inline std::size_t HashOf(T const &v, Others const &...others) {
+    return HashCombine(HashOf(v), HashOf(others...));
+}
+
+template<>
+inline std::size_t HashOf(SimTK::Matrix const &matrix) {
+    std::size_t hash = HashOf(matrix.nrow(), matrix.ncol());
+    for (int r = 0; r < matrix.nrow(); ++r) {
+        for (int c = 0; c < matrix.ncol(); ++c) {
+            hash = HashCombine(hash, matrix.row(r)[c]);
+        }
+    }
+    return hash;
+}
+
+} // namespace OpenSim
+
+template <>
+struct std::hash<SmoothSegmentedFunctionParameters> final
+{
+    size_t operator()(const SmoothSegmentedFunctionParameters& params) const {
+        return HashOf(
+            params._mX, params._mY,
+            params._x0, params._x1,
+            params._y0, params._y1,
+            params._dydx0, params._dydx1,
+            params._computeIntegral,
+            params._intx0x1
+        );
+    }
+};
+
+//=============================================================================
+// SMOOTHSEGMENTEDFUNCTION DATA
+//=============================================================================
+
+namespace OpenSim {
+struct SmoothSegmentedFunctionData {
+        SmoothSegmentedFunctionData(
```

Comment:
Why not define the implementation right here? The only reason to split the declaration/definition is in case the definition uses a symbol that's defined between the two (I can't see one?).